server/src/com/thoughtworks/go/config/GoPartialConfig.java

```diff
@@ -91,32 +76,89 @@ public void onFailedPartialConfig(ConfigRepoConfig repoConfig, Exception ex) {
     @Override
     public synchronized void onSuccessPartialConfig(ConfigRepoConfig repoConfig, PartialConfig newPart) {
         String fingerprint = repoConfig.getMaterialConfig().getFingerprint();
-        if(this.configWatchList.hasConfigRepoWithFingerprint(fingerprint))
-        {
+        if (this.configWatchList.hasConfigRepoWithFingerprint(fingerprint)) {
             //TODO maybe validate new part without context of other partials or main config
 
-            // put latest valid
-            fingerprintToLatestValidConfigMap.put(fingerprint, newPart);
+            // put latest known
+            cachedGoPartials.addOrUpdate(fingerprint, newPart);
+            if (updateConfig(newPart, fingerprint)) {
+                cachedGoPartials.markAsValid(fingerprint, newPart);
+            }
+        }
+    }
 
-            notifyListeners();
+    private boolean updateConfig(final PartialConfig newPart, final String fingerprint) {
+        try {
+            goConfigService.updateConfig(new UpdateConfigCommand() {
+                @Override
+                public CruiseConfig update(CruiseConfig cruiseConfig) throws Exception {
+                    if (newPart != null && fingerprint != null) {
+                        cruiseConfig.getPartials().remove(findMatchingPartial(cruiseConfig, fingerprint));
+                        cruiseConfig.getPartials().add(new Cloner().deepClone(newPart));
+
+                        for (PartialConfig partial : cruiseConfig.getPartials()) {
+                            for (EnvironmentConfig environmentConfig : partial.getEnvironments()) {
+                                if (!cruiseConfig.getEnvironments().hasEnvironmentNamed(environmentConfig.name())) {
+                                    cruiseConfig.addEnvironment(new BasicEnvironmentConfig(environmentConfig.name()));
```

Comment:
Yes, the idea was to add such environments and groups to main xml and save so that it will show up on configXml tab as well as pipelines tab for edit, thus avoiding us to have something like `UIConfigOrigin`.